Redesigned color picker widget

bennothommo opened the pull request:
This PR introduces a newly redesigned color picker widget, built on top of the [Pickr UI widget](https://github.com/Simonwep/pickr) as opposed to Spectrum, which has not seen a release [since 2016](https://github.com/bgrins/spectrum/releases/tag/1.8.0).

![image](https://user-images.githubusercontent.com/15900351/134927135-63387bde-71e8-4399-ba39-f2f30c512ead.png)

### Benefits

- More clean and streamlined UX - available colors are now shown within the popover as opposed to spanning across the entire row at all times. The UI is very similar to VSCode's color picker, which should hopefully encourage familiarity.
- Text input next to the color picker allows full copy+paste capabilities, as well as simply typing out the color code.
- Supports touch screen devices.
- Full accessibility and internationalisation support.
- Full support for other color formats (HEX / RGB / HSL / CMYK) - this PR would replace https://github.com/wintercms/winter/pull/105 and fix https://github.com/wintercms/winter/issues/94
- Will also fix https://github.com/wintercms/winter/issues/321
- Will properly clean up and dispose of the widget as needed.

### New config options

Option | Description
------------- | -------------
**formats** | Specifies the color format(s) to store. Can be a string or an array of values `cmyk`, `hex`, `hsl` and `rgb`. Specifying `all` as a string will allow all formats. Defaults to `hex`.
**allowCustom** | If `false`, only the colors specified in `availableColors` will be available for selection. The color picker palette will be disabled. Default `true`.

### Notes

- Given that we've dropped support for IE11 in the Backend, I've used some ES6 code (just arrow functions, really) to resolve some scoping issues in the JS, and the picker itself uses CSS Grid Layout. Both features, AFAICT, have been supported by all browser versions within the last 4-5 years, so I don't think we're leaving anyone behind with this.
- This should be fully compatible with values from the old color picker, since they would've been stored as hex values. I also believe that it should be able to convert any value to another format if someone opts to change the format in the config, but I have not tested this yet.
- I'll post up a video of this tomorrow, but feel free to have a play around with it on Gitpod in the mean time. Go to **Settings > Customize back-end > Colors** to get a couple of color picker widgets to play with.

<a href="https://gitpod.io/#https://github.com/wintercms/winter/pull/324"><img src="https://gitpod.io/button/open-in-gitpod.svg"/></a>

bennothommo committed:
Replace Spectrum with Pickr

Link: https://github.com/Simonwep/pickr

bennothommo committed:
Convert Spectrum in widget to corresponding Pickr functionality

bennothommo committed:
WIP

bennothommo committed:
Further colorpicker WIP

- Added ability to allow or disallow custom colors
- Made picker more usable and match up to the input
- Picker shows when input is focused
- Added additional format support

bennothommo committed:
Finalising new color picker functionality

bennothommo committed:
Tweaks to allow Enter key to select, but not submit the whole form

bennothommo committed:
Remove debug code

LukeTowers commented:
Something seems off about the font, is that the default font? Maybe we should use a mono space font instead

bennothommo commented:
@LukeTowers yeah it's the default body font. I'll change it to monospace.

bennothommo committed:
Use a monospaced font for preview color value

LukeTowers commented:
Could we also provide `format: any` to allow the user to pick which format they want to use?

bennothommo committed:
Revert codeeditor.css

bennothommo committed:
Remove old template code

bennothommo committed:
Add lang strings for ARIA labels in color picker

bennothommo committed:
Allow for multiple formats and format selection

bennothommo commented:
@LukeTowers multiple format support has been added.

bennothommo committed:
Add additional width to input for larger formats

bennothommo committed:
Add ability to define default swatches for all color pickers.

If available colors are specified on a particular color picker, those will be used instead.

bennothommo committed:
Merge remote-tracking branch 'origin/develop' into wip/colorpicker-rewrite

LukeTowers commented:
Is there a docs PR for the changes?

bennothommo committed:
Fix color picker handling in preview mode

bennothommo commented:
@LukeTowers not yet - I'll put one through shortly. Any further thoughts on this, as it is ready to merge?

LukeTowers commented on modules/backend/formwidgets/colorpicker/partials/_colorpicker.htm:
```diff
@@ -1,42 +1,33 @@
+    <?php if ($allowCustom): ?>data-allow-custom="<?= $allowCustom ?>"<?php endif ?>
+    data-data-locker="#<?= $this->getId('input') ?>"
+    <?php if ($readOnly || $disabled || $this->previewMode): ?>data-disabled="true"<?php endif; ?>
+    <?= $this->formField->getAttributes() ?>>
```
```suggestion
    <?= $this->formField->getAttributes() ?>
>
```

LukeTowers committed:
Update modules/backend/formwidgets/colorpicker/partials/_colorpicker.htm

LukeTowers commented on modules/backend/formwidgets/colorpicker/assets/js/colorpicker.js:
```diff
@@ -9,102 +9,399 @@
 
     ColorPicker.DEFAULTS = {
         showAlpha: false,
+        format: 'hexa',
```
Shouldn't this default to the same default as the PHP?

bennothommo replied:
Good spot.

LukeTowers commented on modules/backend/models/BrandSetting.php:
```diff
@@ -79,6 +80,68 @@ public function initSettingsData()
+            [
+                'color' => '#7f8c8d',
+            ],
+        ]);
```
Why does this format have to be an array of arrays instead of just an array of values?

bennothommo replied:
So it works in the repeater widget :)

bennothommo committed:
Fix default format value

Mrkbingham commented on modules/backend/formwidgets/ColorPicker.php:
```diff
@@ -91,13 +99,14 @@ public function render()
+        $this->vars['allowCustom'] = (bool) $this->allowCustom;
+        $this->vars['showAlpha'] = (bool) $this->showAlpha;
+        $this->vars['readOnly'] = (bool) $this->readOnly;
+        $this->vars['disabled'] = (bool) $this->disabled;
```
I've noticed when setting $field->disabled on the old colorpicker in a model's filterFields function it doesn't inherit the disabled state.  Is that something that has been / could be fixed in this update?

bennothommo replied:
I believe this is working, but will need to double check.
